plaidml/bridge/openvino/ops/proposal.cpp

```diff
@@ -0,0 +1,369 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "plaidml_ops.hpp"
+#include "plaidml_util.hpp"
+
+#include "ngraph/opsets/opset4.hpp"
+
+#include "plaidml/op/op.h"
+
+using namespace plaidml;  // NOLINT[build/namespaces]
+using ngraph::opset4::Proposal;
+
+namespace {
+edsl::Tensor generate_anchors(const ngraph::op::ProposalAttrs& attrs,  //
+                              const int64_t num_anchors,               //
+                              const float coordinates_offset,          //
+                              DType dtype) {
+  auto base_size = attrs.base_size;
+  auto ratios = attrs.ratio;
+  auto num_ratios = ratios.size();
+  auto scales = attrs.scale;
+  auto num_scales = scales.size();
+  IE_ASSERT(num_anchors == num_ratios * num_scales);
+
+  auto round_ratios = attrs.framework != "tensorflow";
+  auto shift_anchors = attrs.framework == "tensorflow";
+
+  float base_area = base_size * base_size;
+  float center = 0.5f * (base_size - coordinates_offset);
+
+  std::vector<float> anchor_coordinates_vec;
+  for (auto ratio : ratios) {
+    float ratio_w;
+    float ratio_h;
+    if (round_ratios) {
+      ratio_w = std::roundf(std::sqrt(base_area / ratio));
+      ratio_h = std::roundf(ratio_w * ratio);
+    } else {
+      ratio_w = std::sqrt(base_area / ratio);
+      ratio_h = ratio_w * ratio;
+    }
+    for (auto scale : scales) {
+      // construct anchor tensor: [num_anchors, 4]
+      auto scale_w = (ratio_w * scale - coordinates_offset) * 0.5f;
+      auto scale_h = (ratio_h * scale - coordinates_offset) * 0.5f;
+      anchor_coordinates_vec.push_back(-scale_w);
+      anchor_coordinates_vec.push_back(-scale_h);
+      anchor_coordinates_vec.push_back(scale_w);
+      anchor_coordinates_vec.push_back(scale_h);
+    }
+  }
+  TensorShape anchor_shape(DType::FLOAT32, {num_anchors, 4});
+  Buffer anchor_buffer(anchor_shape);
+  anchor_buffer.copy_from(anchor_coordinates_vec.data());
```

Comment:
I wonder if we can use `index` to generate a tensor instead of using `copy_from` function. Generally we should avoid using this function.

Reply:
For this part, it seems cannot avoid `copy_from`, as we need to generate `anchor tensor` from two given c++ vector, in which all values are random or set by user, not continuous int in `index`.